crates/flipperzero/src/furi/string.rs

```diff
@@ -0,0 +1,259 @@
+//! String primitives built around `FuriString`.
+
+use core::{
+    cmp::Ordering,
+    convert::Infallible,
+    ffi::{c_char, CStr},
+    fmt, ops,
+};
+
+#[cfg(feature = "alloc")]
+use alloc::ffi::CString;
+
+use flipperzero_sys as sys;
+
+/// A Furi string.
+///
+/// This is similar to Rust's [`String`], but used inside the Flipper Zero SDK.
+#[derive(Eq)]
+pub struct String(*mut sys::FuriString);
+
+impl Drop for String {
+    fn drop(&mut self) {
+        unsafe { sys::furi_string_free(self.0) };
+    }
+}
+
+// Implementations matching `std::string::String`.
+impl String {
+    /// Creates a new empty `String`.
+    pub fn new() -> Self {
+        String(unsafe { sys::furi_string_alloc() })
+    }
+
+    /// Creates a new empty `String` with at least the specified capacity.
+    pub fn with_capacity(capacity: usize) -> Self {
+        let mut s = Self::new();
+        s.reserve(capacity);
+        s
+    }
+
+    /// Extracts a string slice containing the entire `String`.
+    #[inline]
+    #[must_use]
+    pub fn as_str(&self) -> &Str {
+        Str::from_raw(self.0)
+    }
+
+    /// Appends a given `CStr` onto the end of this `String`.
+    pub fn push_cstr(&mut self, string: &CStr) {
+        unsafe { sys::furi_string_cat_str(self.0, string.as_ptr()) }
+    }
+
+    /// Appends a given string slice onto the end of this `String`.
+    pub fn push_str(&mut self, string: &str) {
+        self.reserve(string.len());
+        for ch in string.chars() {
+            self.push(ch);
+        }
+    }
+
+    /// Reserves capacity for at least `additional` bytes more than the current length.
+    pub fn reserve(&mut self, additional: usize) {
+        unsafe { sys::furi_string_reserve(self.0, self.len() + additional) };
+    }
+
+    /// Appends the given [`char`] to the end of this `String`.
+    pub fn push(&mut self, ch: char) {
+        match ch.len_utf8() {
+            1 => unsafe { sys::furi_string_push_back(self.0, ch as c_char) },
+            _ => unsafe { sys::furi_string_utf8_push(self.0, ch as u32) },
+        }
+    }
+
+    /// Shortens this `String` to the specified length.
+    ///
+    /// If `new_len` is greater than the string's current length, this has no effect.
+    pub fn truncate(&mut self, new_len: usize) {
+        if new_len < self.len() {
+            unsafe { sys::furi_string_right(self.0, new_len) };
```

Comment:
Well that's useful to know, because the documentation for `furi_string_right` strongly suggests they are the other way round:
> Return the string right truncated from the 'index' position to the last position